include/exadg/incompressible_navier_stokes/spatial_discretization/turbulence_model.cpp

```diff
@@ -79,18 +101,40 @@ TurbulenceModel<dim, Number>::cell_loop_set_coefficients(
     integrator.evaluate(dealii::EvaluationFlags::gradients);
 
     // get filter width for this cell
-    scalar filter_width = integrator.read_cell_data(this->filter_width_vector);
+    if(use_turbulence_model)
+    {
+      filter_width = integrator.read_cell_data(this->filter_width_vector);
+    }
 
     // loop over all quadrature points
     for(unsigned int q = 0; q < integrator.n_q_points; ++q)
     {
-      scalar viscosity = dealii::make_vectorized_array<Number>(turb_model_data.kinematic_viscosity);
-
-      // calculate velocity gradient
-      tensor velocity_gradient = integrator.get_gradient(q);
-
-      add_turbulent_viscosity(viscosity, filter_width, velocity_gradient, turb_model_data.constant);
+      // calculate needed quantities for this cell
+      tensor velocity_gradient           = integrator.get_gradient(q);
+      tensor symmetric_velocity_gradient = dealii::make_vectorized_array<Number>(0.5) *
+                                           (velocity_gradient + transpose(velocity_gradient));
+      scalar shear_rate_squared =
+        dealii::make_vectorized_array<Number>(2.0) *
+        scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient);
```

Comment:
These lines of code seem to appear multiple times (see also face-loop and boundary-face-loop). It could make sense to define free (inlined) functions to avoid duplicated code. Also, it does not seem to be a modular design, i.e. implementing here everything that is needed by at least one of the models. The alternative is to compute here only the velocity gradient, and then do all the other calculations inside the respective viscosity model. Somewhat more floating point operations, but will this ever show up in terms of computational efficiency?

The code would be even more modular when writing separate cell/face/boundary-face loops for every type of viscosity model. Then, the models can be plugged together in a very modular way on the level of `IncNS::SpatialOperatorBase`. The downside would be multiple implementations of `integrator.reinit()/read_dof_values()/evaluate()`, which is a general problem in ExaDG (and would require some higher-level matrix-free functionality than currently available via deal.II). It could make sense to discuss these aspects in person, maybe together with @kronbichler regarding performance aspects.